<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-04-17" />

<title>h-magma</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">multi_omics_neuroscience</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/han16/multi_omics_neuroscience">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">h-magma</h1>
<h4 class="date">2025-04-17</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-04-23
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>multi_omics_neuroscience/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20250206code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20250206)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20250206code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20250206)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomhan16multiomicsneurosciencetreed39784c6df797a977a367b66d485273844dcec11targetblankd39784ca">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/han16/multi_omics_neuroscience/tree/d39784c6df797a977a367b66d485273844dcec11" target="_blank">d39784c</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomhan16multiomicsneurosciencetreed39784c6df797a977a367b66d485273844dcec11targetblankd39784ca"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/han16/multi_omics_neuroscience/tree/d39784c6df797a977a367b66d485273844dcec11" target="_blank">d39784c</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory

Unstaged changes:
    Modified:   analysis/hmagma_nature_protocol.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/hmagma_nature_protocol.Rmd</code>) and HTML
(<code>docs/hmagma_nature_protocol.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/han16/multi_omics_neuroscience/blob/d39784c6df797a977a367b66d485273844dcec11/analysis/hmagma_nature_protocol.Rmd" target="_blank">d39784c</a>
</td>
<td>
han
</td>
<td>
2025-04-21
</td>
<td>
4/21/2025
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/han16/multi_omics_neuroscience/d39784c6df797a977a367b66d485273844dcec11/docs/hmagma_nature_protocol.html" target="_blank">d39784c</a>
</td>
<td>
han
</td>
<td>
2025-04-21
</td>
<td>
4/21/2025
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<ul>
<li>the codes are from nature protocol paper <a
href="https://www.nature.com/articles/s41596-022-00745-z">Annotating
genetic variants to target genes using H-MAGMA</a>.</li>
</ul>
<pre class="r"><code>options(stringsAsFactors=F)
library(GenomicRanges)
library(biomaRt)
library(dplyr)</code></pre>
<div
id="create-genomicranges-granges-objects-for-exon-and-promoter-coordinates-timing-5-min"
class="section level2">
<h2>Create GenomicRanges (GRanges) objects for exon and promoter
coordinates ● Timing ~5 min</h2>
<pre class="r"><code># 3 Read in exonic and promoter coordinate ﬁles by entering the following commands
path=&quot;C:\\han\\Dataset\\20250221_PMID_32152537_hmagma\\6382668\\HMAGMA_Protocol_v1.02\\HMAGMA_Protocol\\&quot;
path2=&quot;C:\\han\\Projects\\2025_01_Improve_MAGMA\\&quot;

exon &lt;- read.table(paste(path, &quot;Gencode26_exon.bed&quot;, sep=&quot;&quot;))
exon$V1 &lt;- sub(&quot;^&quot;, &quot;chr&quot;, exon$V1)
promoter &lt;- read.table(paste(path, &quot;Gencode26_promoter.bed&quot;, sep=&quot;&quot;))
promoter$V1 &lt;- sub(&quot;^&quot;, &quot;chr&quot;, promoter$V1)

# 4 Create a GRanges object for exon and promoter deﬁnitions.
exonranges &lt;- GRanges(exon[,1],IRanges(exon[,2],exon[,3]),gene=exon[,4])
promoterranges &lt;- GRanges(promoter[,1], IRanges(promoter[,2], promoter[,3]), gene=promoter[,4])

# 5 Save exon and promoter GRanges objects as a .rda ﬁle for future use.
#save(exonranges, promoterranges, file=&quot;C:\\han\\Projects\\2025_01_Improve_MAGMA\\exon_promoranges.rda&quot;)</code></pre>
</div>
<div id="generate-a-granges-object-for-snp-coordinates-timing-5-min"
class="section level2">
<h2>Generate a GRanges object for SNP coordinates ● Timing ~5 min</h2>
<pre class="r"><code>#Generate a GRanges object for SNP coordinates ● Timing ~5 min
#6 Read in the SNP annotation ﬁle by using the following command.
snps &lt;- read.table(paste(path, &quot;EUR.bim.gz&quot;, sep=&quot;&quot;))
snps &lt;- snps[,c(1,2,4)]
colnames(snps) &lt;- c(&quot;chr&quot;,&quot;SNP&quot;,&quot;Position&quot;)
snps$chr &lt;- sub(&quot;^&quot;, &quot;chr&quot;, snps$chr)


#7 Create a GRanges object for SNP annotations.
snps &lt;- GRanges(snps$chr, IRanges(snps$Position, snps$Position), rsid=
snps$SNP)
#8 Save the SNP GRanges object as a .rda ﬁle for future use.
#save(snps,file=paste(path2, &quot;snps.rda&quot;, sep=&quot;&quot;))</code></pre>
</div>
<div
id="assign-snps-to-genes-by-overlapping-snps-with-exons-and-promoters-timing-1015-min"
class="section level2">
<h2>Assign SNPs to genes by overlapping SNPs with exons and promoters ●
Timing 10–15 min</h2>
<pre class="r"><code>#9 Identify genes that map to SNPs residing in exons by overlapping the GRanges object for SNPs with
#the GRanges object for exons.
olap &lt;- findOverlaps(snps,exonranges)
snpexon &lt;- snps[queryHits(olap)]
mcols(snpexon) &lt;- cbind(mcols(snpexon), mcols(exonranges[subjectHits(olap)]))
snpexon &lt;- snpexon[seqnames(snpexon)!=&quot;chrX&quot;]


#10 Overlap the GRanges object for promoters with the GRanges object for SNPs. Similar to the code
#above, this command will identify genes that map to SNPs residing in promoter regions.
olap &lt;- findOverlaps(snps,promoterranges)
snpro &lt;- snps[queryHits(olap)]
mcols(snpro) &lt;- cbind(mcols(snpro),mcols(promoterranges[subjectHits(olap)]))
snpro &lt;- snpro[seqnames(snpro)!=&quot;chrX&quot;]

#11 Save SNPs overlapping with exons and promoters as a .rda ﬁle.
#save(snpro, snpexon, file=paste(path2, &quot;snp_locating_in_exon_promoter_transcript_level.rda&quot;, sep=&quot;&quot;))</code></pre>
</div>
<div
id="assign-unmapped-snps-to-genes-on-the-basis-of-hi-c-interaction-data-timing"
class="section level2">
<h2>Assign unmapped SNPs to genes on the basis of Hi-C interaction data
● Timing</h2>
<p>30–45 min</p>
<pre class="r"><code># 12 After identiﬁcation of SNPs that reside in exons and/or promoters, there will be a subset of SNPs
# that do not overlap with either (e.g., intergenic or intronic SNPs). In the following steps, we match
# those SNPs to cognate genes on the basis of Hi-C interaction data. Identify unmapped SNPs by
#using the code below.

snpranges &lt;- snps[!(snps$rsid %in% snpexon$rsid),]
snpranges &lt;- snpranges[!(snpranges$rsid %in% snpro$rsid),]


#13 Save unmapped SNPs as a .rda ﬁle.
#save(snpranges, file=paste(path2, &quot;non_exonic_promoter_snp.rda&quot;, sep=&quot;&quot;))

#14 Read in adult brain Hi-C data. The ﬁrst three columns represent genomic coordinates [chr, start,
#end] of loop anchor 1, and the last three columns represent genomic coordinates of loop anchor 2.
#CRITICAL STEP Although we use a Hi-C dataset from the adult human brain for this protocol,
#Hi-C datasets from the tissue or cell type associated with the disease/trait of interest can be used
#instead.
hic &lt;- read.table(&quot;adultbrain_hic.bedpe&quot;, header=T)


#15 Restructure Hi-C data to account for both anchors 1 and 2 by using the command below.
hic.int1 &lt;- hic[,1:6]
hic.int2 &lt;- hic[,c(4:6,1:3)]
colnames(hic.int1) = colnames(hic.int2) = c(&quot;chrom1&quot;, &quot;start1&quot;, &quot;end1&quot;, &quot;chrom2&quot;, &quot;start2&quot;, &quot;end2&quot;)
hic.comb &lt;- rbind(hic.int1, hic.int2)

#16 Generate a GRanges object for the adult brain Hi-C data by using the command below.
hicranges &lt;- GRanges(hic.comb$chrom1, IRanges(as.numeric(hic.comb$start1), as.numeric(hic.comb$end1)), int1=hic.comb$start2,int2=hic.comb$end2)

#17 Identify promoter-anchored interactions by overlapping loop anchor 1 with promoters.
olap &lt;- findOverlaps(hicranges,promoterranges)
generanges &lt;- hicranges[queryHits(olap)]
mcols(generanges) &lt;- cbind(mcols(hicranges[queryHits(olap)]), mcols(promoterranges[subjectHits(olap)]))


#18 Reverse the order of the GRanges object from Step 17. The generanges object created in Step 17
#is in a format of loop anchor 1, loop anchor 2, gene name. We reverse the order of loop
#anchors so that the resulting genebed object has a format of loop anchor 2, loop anchor 1,
#gene name.
genebed &lt;- data.frame(chr=seqnames(generanges), snp.start=generanges$int1, snp.end=generanges$int2, gene.start=start(generanges), gene.end=start(generanges)+width(generanges)-1, ensg=generanges$gene)
genebed &lt;- unique(genebed)


#19 Create a GRanges object from Step 18.
genesnpranges &lt;- GRanges(genebed$chr, IRanges(genebed$snp.start,genebed$snp.end), ensg=genebed$ensg)



#20 Overlap unmapped SNPs from Step 12 with loop anchor 2 from Step 19. This step assigns
#SNPs (located at loop anchor 2) to the genes with which they interact (located at loop
#anchor 1).
olap &lt;- findOverlaps(snpranges,genesnpranges)
snpint &lt;- snpranges[queryHits(olap)]
mcols(snpint) &lt;- cbind(mcols(snpranges[queryHits(olap)]), mcols(genesnpranges[subjectHits(olap)]))
#save(snpint, file=paste(path2, &quot;Hi-C_transcript_interacting_snp.rda&quot;, sep=&quot;&quot;))


#21 Integrate SNP-gene relationships derived from exons, promoters and Hi-C interaction data.
load(paste(path2, &quot;Hi-C_transcript_interacting_snp.rda&quot;, sep=&quot;&quot;))
load(paste(path2, &quot;snp_locating_in_exon_promoter_transcript_level.rda&quot;, sep=&quot;&quot;))
snpdat &lt;- data.frame(chr=seqnames(snpint),
bp=start(snpint), rsid=snpint$rsid, ensg=snpint$ensg)
snpromat &lt;- unique(data.frame(rsid=snpro$rsid, ensg=snpro$gene))
snpexonmat &lt;- unique(data.frame(rsid=snpexon$rsid, ensg=snpexon$gene))
snpcomb &lt;- unique(rbind(snpdat[,3:4], snpromat, snpexonmat))
#save(snpcomb, file=paste(path2, &quot;SNP_to_transcript_comb.rda&quot;, sep=&quot;&quot;))</code></pre>
</div>
<div
id="create-the-h-magmacompatible-variant-gene-annotation-ﬁle-timing-1520-min"
class="section level2">
<h2>Create the H-MAGMA–compatible variant-gene annotation ﬁle ● Timing
15–20 min</h2>
<pre class="r"><code>#22 Aggregate SNP-gene relationships to generate the variant-gene annotation ﬁle compatible
#with MAGMA.
snpagg &lt;- aggregate(snpcomb, list(snpcomb$ensg), unique)


#23 Read in the gene deﬁnition ﬁle.
genedef &lt;- read.table(paste(path, &quot;Gencode26_gene.bed&quot;, sep=&quot;&quot;))
colnames(genedef) &lt;- c(&quot;chr&quot;, &quot;start&quot;, &quot;end&quot;, &quot;ensg&quot;)


#24 Create an index column from the gene deﬁnition ﬁle consisting of gene chromosomal location, start
#and end.
genedef &lt;- genedef [grep(&quot;chr&quot;, genedef$chr),]
genedef$chr &lt;- unlist(lapply(strsplit(genedef$chr, &quot;chr&quot;), &#39;[[&#39;, 2))
genedef$index &lt;- paste(genedef$chr, genedef$start,
genedef$end, sep=&quot;:&quot;)


#25 Attach the index column from Step 24 to the variant-gene annotation ﬁle from Step 22.
snpagg$index &lt;- genedef[match(snpagg$ensg, genedef$ensg),&quot;index&quot;]


#26 Remove any missing values from the variant-gene annotation ﬁle.
snpagg &lt;- snpagg[!is.na(snpagg$index),]

#27 Subset gene, gene location and SNPs from the variant-gene annotation ﬁle.
snpannot &lt;- snpagg[,c(&quot;ensg&quot;, &quot;index&quot;, &quot;rsid&quot;)]

#28 Save the variant-gene annotation ﬁle in an executable format.
writable &lt;- format(snpannot)
write.table(writable, file=paste(path2, &quot;SNP_aggregate_transcript.txt&quot;, sep=&quot;&quot;), quote=F, row.names=F, col.names=F, sep=&quot;\t&quot;)

#system(paste0(&quot;sed -e `s/, /\t/g` &lt; SNP_aggregate_transcript&quot;,&quot;.txt &gt;&quot;,&quot;Adultbrain.transcript.annot&quot;))
system(&quot;sed -e &#39;s/, /\\t/g&#39; &lt; SNP_aggregate_transcript.txt &gt; Adultbrain.transcript.annot&quot;)</code></pre>
<pre><code>[1] 2</code></pre>
<pre class="r"><code># R code alternative for the above command from ChatGPT
lines &lt;- readLines(paste(path2, &quot;SNP_aggregate_transcript.txt&quot;, sep=&quot;&quot;))
lines &lt;- gsub(&quot;, &quot;, &quot;\t&quot;, lines)
#writeLines(lines, paste(path2, &quot;Adultbrain.transcript.annot&quot;, sep=&quot;&quot;))</code></pre>
</div>
<div id="run-h-magma-timing-3050-min" class="section level2">
<h2>Run H-MAGMA ● Timing 30–50 min</h2>
<pre class="r"><code># run magma in the magma folder 
magma --bfile C:\\han\\Dataset\\20250221_PMID_32152537_hmagma\\g1000_eur\\g1000_eur 
--pval C:\\han\\Dataset\\20250221_PMID_32152537_hmagma\\6382668\\HMAGMA_Protocol_v1.02\\HMAGMA_Protocol\\PD.summary.stat.txt use=rs, p ncol=N 
--gene-annot C:\\han\\Projects\\2025_01_Improve_MAGMA\\Adultbrain.transcript.annot  
--out C:\\han\\Projects\\2025_01_Improve_MAGMA\\PD_GWAS\\</code></pre>
<pre class="r"><code>#31 Retrieve the number of genes associated with PD at different thresholds after multiple corrections
#by running the following commands. Genes can be stratiﬁed into either protein-coding or
#noncoding genes

options(stringsAsFactors=F)
#setwd (&quot;~/work/&quot;)
load(paste(path, &quot;geneAnno_allgenes.rda&quot;, sep=&quot;&quot;))
backgroundset&lt;-unique(geneAnno1[geneAnno1$gene_biotype==&quot;protein_coding&quot;, &quot;ensembl_gene_id&quot;])
diseasename &lt;- c(&quot;PD&quot;)
fdrdisease &lt;- c()
diseasemat &lt;- read.table(paste(path2, &quot;PD_GWAS\\PD_GWAS.genes.out.txt&quot;, sep=&quot;&quot;), header=T)
diseasemat &lt;- diseasemat[diseasemat$GENE %in% backgroundset,]
diseasemat$FDR &lt;- p.adjust(diseasemat$P, &quot;BH&quot;)
backgroundensg &lt;- diseasemat$GENE
queryensg0 &lt;- diseasemat[diseasemat$FDR&lt;0.1, &quot;GENE&quot;]
queryensg1 &lt;- diseasemat[diseasemat$FDR&lt;0.05, &quot;GENE&quot;]
queryensg2 &lt;- diseasemat[diseasemat$FDR&lt;0.01, &quot;GENE&quot;]
fdrgene &lt;- c(diseasename, length(queryensg0), length(queryensg1),
length(queryensg2))
fdrdisease &lt;- rbind(fdrdisease, fdrgene)
colnames(fdrdisease) &lt;- c(&quot;disease&quot;, &quot;FDR&lt;0.1&quot;, &quot;FDR&lt;0.05&quot;, &quot;FDR&lt;0.01&quot;)
fdrdisease &lt;- data.frame(fdrdisease)
#write.csv(fdrdisease, file=paste(path2, &quot;PD_GWAS\\PD.genes.csv&quot;,sep=&quot;&quot;), col.names=T, row.names=F,
#sep=&quot;\t&quot;, quote=F)

#The commands above provide the number of protein-coding genes at different thresholds. To retrieve the number of all genes at the same threshold, simply comment out the command line as displayed below.
#diseasemat&lt;-diseasemat[diseasemat$GENE%in% backgroundset,]. This will treat the command line as a comment and prevent it from being included in the rest of the analysis, thereby generating a list of all genes rather than just protein-coding genes.</code></pre>
<pre class="r"><code>#32 Retrieve a list of PD risk genes at false discovery rate (FDR) &lt; 0.05 in HGNC symbols by entering
#the command below.
pd.genes&lt;-unique(geneAnno1[match(queryensg1,geneAnno1$ensembl_gene_id), &quot;hgnc_symbol&quot;])</code></pre>
<pre class="r"><code>#33 Generate a bar plot to compare the number of all genes versus protein-coding genes at different
#thresholds by using the commands below.
library(tidyr)
library(ggplot2)
df &lt;- read.csv(paste(path2, &quot;PD_GWAS\\PD.genes.csv&quot;, sep=&quot;&quot;), header = T)
df$Threshold &lt;- row.names(df)
lab &lt;- c(&quot;FDR&lt;0.1&quot;,&quot;FDR&lt;0.05&quot;,&quot;FDR&lt;0.01&quot;)
df_long &lt;- gather(df, key = var, value = value, All_genes,PCG)
ggplot(df_long, aes(x = Disorder, y = value, ﬁll = var)) +
geom_bar(stat = &#39;identity&#39;, position = &#39;dodge&#39;) +
scale_x_discrete(labels = lab)+
geom_text(aes(label=value), vjust=1.6, Color=&quot;black&quot;)+
scale_ﬁll_manual(values = c(&quot;#00AFBB&quot;, &quot;#CC79A7&quot;))</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.0 (2024-04-24 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 10 x64 (build 19045)

Matrix products: default


locale:
[1] LC_COLLATE=English_United States.utf8 
[2] LC_CTYPE=English_United States.utf8   
[3] LC_MONETARY=English_United States.utf8
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
[1] dplyr_1.1.4          biomaRt_2.60.1       GenomicRanges_1.56.2
[4] GenomeInfoDb_1.40.0  IRanges_2.38.0       S4Vectors_0.42.0    
[7] BiocGenerics_0.50.0 

loaded via a namespace (and not attached):
 [1] KEGGREST_1.44.0         xfun_0.43               bslib_0.7.0            
 [4] httr2_1.0.1             Biobase_2.64.0          generics_0.1.3         
 [7] vctrs_0.6.5             tools_4.4.0             curl_5.2.1             
[10] tibble_3.2.1            fansi_1.0.6             AnnotationDbi_1.66.0   
[13] RSQLite_2.3.6           blob_1.2.4              pkgconfig_2.0.3        
[16] dbplyr_2.5.0            lifecycle_1.0.4         GenomeInfoDbData_1.2.12
[19] compiler_4.4.0          stringr_1.5.1           git2r_0.33.0           
[22] Biostrings_2.72.0       progress_1.2.3          httpuv_1.6.15          
[25] htmltools_0.5.8.1       sass_0.4.9              yaml_2.3.8             
[28] later_1.3.2             pillar_1.9.0            crayon_1.5.2           
[31] jquerylib_0.1.4         whisker_0.4.1           cachem_1.0.8           
[34] tidyselect_1.2.1        digest_0.6.35           stringi_1.8.4          
[37] rprojroot_2.0.4         fastmap_1.1.1           cli_3.6.2              
[40] magrittr_2.0.3          utf8_1.2.4              filelock_1.0.3         
[43] prettyunits_1.2.0       UCSC.utils_1.0.0        promises_1.3.0         
[46] rappdirs_0.3.3          bit64_4.0.5             rmarkdown_2.26         
[49] XVector_0.44.0          httr_1.4.7              bit_4.0.5              
[52] workflowr_1.7.1         png_0.1-8               hms_1.1.3              
[55] memoise_2.0.1           evaluate_0.23           knitr_1.46             
[58] BiocFileCache_2.12.0    rlang_1.1.3             Rcpp_1.0.12            
[61] glue_1.7.0              DBI_1.2.3               xml2_1.3.6             
[64] rstudioapi_0.16.0       jsonlite_1.8.8          R6_2.5.1               
[67] fs_1.6.4                zlibbioc_1.50.0        </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
